### DISCUSSION TITLE
postversion script not triggered.

### DISCUSSION THREAD

**Discussion by mmellado** [semantic-release/semantic-release#2944] [Q&A]
I'm using semantic release to automate the release flow of an npm library I'm trying to publish. Things work fine with 1 exception.

My library is written in typescript which means I need to build it before publishing it. This means the published directory is not the source, but a `dist` directory that is created on build prior to me triggering semantic-release.

I've configured `semantic-release/npm` and `semantic-release/git` to focus on the dist directory, and this works fine, the published files are only those in the `dist` directory. However, to make sure that everything is up to date, after semantic-release/npm updates package.json, I want to also update the version in the root of the project. I read that this can be accomplished using the `postversion` hook via npm scripts. However, this doesn't seem to be happening, or at least, at the end of the flow, it's not attempting to make a commit with the updated CHANGELOG and package.json files.

This is the repository https://github.com/velupjs/velup/tree/v1.0.0-alpha.1 (alpha branch is where I'm doing this as the library is not ready for release yet, specially as I'm testing this release flow).

Any ideas on what I'm missing? My main goal is to make sure that the version is updates on both `package.json` and `dist/package.json`, but only the contents of `dist` are published to npm

**Comment by mmellado** [comment]
As a note, I've fixed my issue using `@semantic-release/exec`. However, I'm still curious to understand why it wasn't working when using only the `postversion` script

**Comment by seebeen** [comment]
You do not need to update the version in package.json. `@semantic-release/npm` package does that automatically for you.

For a fully implemented typescript package release flow you can view the release config on the [@semantic-release/wordpress](/semantic-release/wordpress) package. Feel free to copy everything you need from there.

**Comment by mmellado** [comment]
This part is clear to me. My package.json version does get updated but only on the compiled version, not on the source, because I’m pointing the root dir to be dist. Since only the package.json in dist is updated, I need to still copy the new version to the main package.json. In the same way, changelog is only updated in dist and not in the source so I also need to copy this file from dist to source after semantic release updates it. The implementation in the branch I posted above is following a recipe from the documentation of semantic-release/npm to copy the files using the postversion hook, but the hook is not triggered

--

Marcos Mellado

**Reply by seebeen** [reply]
Why do you have two package.json files? 🤔

**Reply by mmellado** [reply]
I don't....

There is the source, what you see on github as the repository. Then there is the compiled version, which is the files that get published to npm. Since I'm not publishing the whole source, but only the compiled library (aka, the files that get outputted to the dist directory), the package.json file needs to be available in the dist directory, so as part of the build, it is copied there.

After the library is built, then I try to publish it. Thats when I run semantic-release with `semantic-release/npm`  and `semantic-release/git` (and others) to update the version in package.json and the release changelog. Now, since I'm only publishing the dist directory, I use the following as part of my configuration:

```
[
      "@semantic-release/npm",
      {
        "pkgRoot": "dist",
        "tarballDir": "dist"
      }
    ],
    [
      "@semantic-release/git",
      {
        "assets": [
          "package.json",
          "CHANGELOG.md"
        ],
        "message": "chore(release): ${nextRelease.version} [skip ci]\n\n${nextRelease.notes}"
      }
    ]
```

This would make the npm plugin trigger the dist directory (the compiled version) and not the source, meaning it updates the package.json file in `dist/package.json`. With the git extension later trying to commit the updated files to the repo.

The problem is, since the contents of dist are compiled code, these are not committed to the repository, so I need to make sure that the changelog is preserved and that the version of the root (source) package are updated as well.

If you check the README file of the npm plugin, at the end there is an example that covers my usecase: https://github.com/semantic-release/npm#examples with the README stating the following:

> When publishing from a sub-directory with the pkgRoot option, the package.json and npm-shrinkwrap.json updated with the new version can be moved to another directory with a postversion. For example with the [@semantic-release/git](https://github.com/semantic-release/git) plugin

This is exactly my use-case. I'm publishing from a sub-rirectory and need to copy the package.json to another directory. However, the execution didn't work as the README explained, as even when I had a `posversion` hook in my package.json's scripts, the files were not copied to the root directory, and hence, not updated in the repository.

Like I mentioned in my previous comment, I've been able to fix my usecase by introducing exec as follows:

```
[
      "@semantic-release/npm",
      {
        "pkgRoot": "dist",
        "tarballDir": "dist"
      }
    ],
    [
      "@semantic-release/exec",
      {
        "prepareCmd": "ts-node scripts/post-version.ts ${nextRelease.version}"
      }
    ],
    [
      "@semantic-release/git",
      {
        "assets": [
          "package.json",
          "CHANGELOG.md"
        ],
        "message": "chore(release): ${nextRelease.version} [skip ci]\n\n${nextRelease.notes}"
      }
    ]
```

with `post-version.ts` simply copying the files I need, but what I'm trying to get from this discussion is an understanding of why the postversion hook was not being triggered (regardless of what it did or why), given the documentation said it should. I'm not sure if it didn't trigger because I did something wrong, or because there is a bug that prevented it from triggering.

**Comment by travi** [comment]
it seems like you are attempting to follow a path that is far more complicated than it sounds like you need. 

first, i would recommend avoiding use of `pkgRoot` unless there are details that i'm not understanding about your need. instead, point `main`/`exports`/etc properties in your `package.json` to the built files in your `dist/` directory and make sure those files are included in your release. we recommend use of the `files` property in your `package.json` to define the included files as an include-list rather than defining a `.npmignore` file. you can trigger your build script using npm directly by defining a `prepack` script, so that it runs before the publish step. 

beyond that, we also recommend against use of the git and changelog plugins unless you _absolutely_ need them because adding commits during the release process _does_ significantly complicate things, even though semantic-release handles adding them as gracefully as is possible. since you are [already using github releases](https://github.com/velupjs/velup/releases), you could prefer publishing release notes there instead of to a `CHANGELOG.md` file. use the npm registry as your source of truth for latest version, like `npm dist-tags ls <your-package-name>` instead of the `package.json` under version control. semantic-release handles updating the `package.json` before publishing and does not need the version to be committed to that file.

**Comment by seebeen** [comment]
What @travi said. You can also use the files directive in `package.json`, and .npmignore to skip tarballing files you don't need.
Again, if you check [@semantic-release/wordpress](/semantic-release/wordpress) plugin, you'll see that only the files in dist are published.

My advice would be to copy the flow from there 1:1 and see if that solves the issues you're having.
